### PR TITLE
Cloud energy collection 

### PR DESCRIPTION
**Pull request by anish-mudaraddi**
Draft adding iriscasttools - a python package to get energy usage metrics from cloud hypervisors and storage nodes.
Collects a variety of information including:
- instantaneous power in Watts
- Os load 
- RAM usage

Python package called by a cron job every 5 minutes on each hypervisor to collect data over time

### CODE REVIEW

**Comment by anish-mudaraddi** [comment]
This tool makes use of `subprocess.popen` to call `ipmitool` to get power information.

There exists other python packages that aim to provide ipmi functionality as a python package

- [python-ipmi](https://github.com/kontron/python-ipmi)
- [hyve-pyipmi](https://github.com/hyvedesignsolutions/hyve-pyipmi)
 
Neither seem to allow us to execute dcmi commands - needed to get power information

Both seem to require a username and password for accessing ipmitool which we don't require here since the command is run locally

So I think the current approach is fine as is

File: .github/workflows/iriscast_package_build.yaml
Comment (DavidFair) [inline .github/workflows/iriscast_package_build.yaml:17]: We should also test against the version of Python shipped in RL8 too

File: cloud-energy-collection/iriscasttools/main.py
Comment (DavidFair) [inline cloud-energy-collection/iriscasttools/main.py:7]: This should be wrapped in a `if __main__` too:
https://stackoverflow.com/questions/419163/what-does-if-name-main-do

Like:
```
if __main__:
   get_iriscast_stats(...)
```

File: cloud-energy-collection/iriscasttools/test/test_main.py
Comment (DavidFair) [inline cloud-energy-collection/iriscasttools/test/test_main.py:47]: We should check that what `get_iriscast_stats` returns was the same as what one of the mock methods returned

Currently you could change it to `return None` and the test would still pass

File: cloud-energy-collection/iriscasttools/test/test_utils.py
Comment (DavidFair) [inline cloud-energy-collection/iriscasttools/test/test_utils.py:125]: Do we have a test for if these paths don't exist? I.e. if someone accidentally runs this on a VM

File: cloud-energy-collection/iriscasttools/test/test_utils.py
Comment (DavidFair) [inline cloud-energy-collection/iriscasttools/test/test_utils.py:201]: Nit: This can simply be `mock_check_ipmi_conn.assert_called_once()`
As it's more Pythonic

File: cloud-energy-collection/iriscasttools/utils.py
Comment (DavidFair) [inline cloud-energy-collection/iriscasttools/utils.py:128]: When regex is involved, we should have line-by-line comments explaining what you are doing, as it can quickly suck up dev time

Ideally, I would push you to simplify with tools like `str.split()` then filter using lists instead of regex. It's more terse, but always way easier to maintain or change long term

File: cloud-energy-collection/iriscasttools/utils.py
Comment (DavidFair) [inline cloud-energy-collection/iriscasttools/utils.py:159]: These should warn to a logger if they fail

File: cloud-energy-collection/iriscasttools/utils.py
Comment (DavidFair) [inline cloud-energy-collection/iriscasttools/utils.py:205]: Ditto

**Review by DavidFair — CHANGES_REQUESTED** [review]

**Comment by anish-mudaraddi** [comment]
can't figure out why the python 3.6 test is failing. Other than that it's ready to review

**Review by DavidFair — COMMENTED** [review]
The pylint error is due to a slightly non-standard layout

If we move it to the following structure:
```
cloud-energy-collection/main.py
cloud-energy-collection/test/*
cloud-energy-collection/iriscasttools/utils.py
cloud-energy-collection/iriscasttools/*.py
```

Then we simply do `cd cloud-energy-collection && pylint --recursive .`  (notice test isn't included to add it twice)

I'm happy for you to call the top level dir to `cloud-energy-collection` to `iriscasttools`. It wouldn't be the only thing with a bespoke name :wink:

**Comment by DavidFair** [comment]
> I'm going to change the structure of the files to this: ...

Just needs one tweak, `main.py` should sit along `requirements.txt`
The idea being, main imports `iriscasttools.utils` when it's running. In the future the implementation will all go into the subpackage whilst "users" simply use main from the outside.

**Comment by anish-mudaraddi** [comment]
> Just needs one tweak, `main.py` should sit along `requirements.txt`

I added `__main__.py` so the users can call `python -m iriscastools <args>` 
I thought that made more sense :)

**Comment by anish-mudaraddi** [comment]
> I prefer the second approach,

it seems better practice to keep tests outside the package 
https://doc.pytest.org/en/latest/explanation/goodpractices.html#tests-outside-application-code

**Comment by DavidFair** [comment]
> > I prefer the second approach,
> 
> it seems better practice to keep tests outside the package https://doc.pytest.org/en/latest/explanation/goodpractices.html#tests-outside-application-code

I'll leave this up to your call, there's pros and cons

Outside:
Pros: You don't build them into the wheel. No chance someone would ever import something from test into the main application
Cons: Have to handle Python paths locally and setting up the CI. Some IDEs struggle

Inside:
Pros: Can run unit tests after installing your wheel if you want to test against a target platform.
Cons: Wheel size / above

Beyond that, it's a developer's call

**Comment by anish-mudaraddi** [comment]
> I'll leave this up to your call, there's pros and cons
I'll leave the tests outside

I'm happy with this - I'll look into setting up wheels once this is merged

File: iriscasttools/iriscasttools/__main__.py
Comment (DavidFair) [inline iriscasttools/iriscasttools/__main__.py:1]: This file should really be called `main` as the main entrypoint

File: iriscasttools/iriscasttools/__main__.py
Comment (DavidFair) [inline iriscasttools/iriscasttools/__main__.py:5]: It would be nice to pre-configure the logger for the module like this:
https://github.com/stfc/SCD-OpenStack-Utils/blob/master/OpenStack-Rabbit-Consumer/entrypoint.py#L9

That way we can easily do something like
`LOG_LEVEL=DEBUG python3 -m iriscasttools` if something isn't working
...

File: iriscasttools/iriscasttools/utils.py
Comment (DavidFair) [inline iriscasttools/iriscasttools/utils.py:38]: Can we add debug logging for what args and kwargs we're about to try before
and something to debug log the return code

File: iriscasttools/iriscasttools/utils.py
Comment (DavidFair) [inline iriscasttools/iriscasttools/utils.py:62]: Don't bother passing default values from above. Let's simplify to simply:

```
@retry(retry_on=(AssertionError,))
```

File: iriscasttools/iriscasttools/utils.py
Comment (DavidFair) [inline iriscasttools/iriscasttools/utils.py:21]: We can capture the logger from the module scope here to remove boilerplate later on:

```
retry_logger: logging.Logger = logger
```

File: iriscasttools/iriscasttools/utils.py
Comment (DavidFair) [inline iriscasttools/iriscasttools/utils.py:73]: Throw an exception instead of assert

Assert implies a programming error in Python
Exception implies a logical / runtime error, such as the OS not having the package

File: iriscasttools/iriscasttools/utils.py
Comment (DavidFair) [inline iriscasttools/iriscasttools/utils.py:86]: Ditto on these throwing if it doesn't make sense to continue throughout this file

File: iriscasttools/iriscasttools/utils.py
Comment (DavidFair) [inline iriscasttools/iriscasttools/utils.py:138]: If `ipmi_raw_power_query` throws this whole block can be simplified by removing the if condition

File: iriscasttools/setup.py
Comment (DavidFair) [inline iriscasttools/setup.py:16]: Nit: >=3.6

File: iriscasttools/test/test_utils.py
Comment (DavidFair) [inline iriscasttools/test/test_utils.py:39]: Nit: do this outside the with context

File: iriscasttools/iriscasttools/utils.py
Comment (anish-mudaraddi) [reply iriscasttools/iriscasttools/utils.py:46]: That's fair 
The reason I did this was so we didn't want miss collecting other metrics if one fails

**Review by DavidFair — APPROVED** [review]
LGTM

**Pull request merged**
